core/src/main/java/io/grpc/Metadata.java

```diff
@@ -396,8 +411,37 @@ private String toStringInternal() {
     private final String name;
     private final byte[] asciiName;
 
+    private static Set<Character> generateValidTChars() {
+      ImmutableSet.Builder<Character> valid = ImmutableSet.builder();
+      valid.add('-');
+      valid.add('_');
+      for (char c = '0'; c <= '9'; c++) {
+        valid.add(c);
+      }
+      for (char c = 'a'; c <= 'z'; c++) {
+        valid.add(c);
+      }
+      return valid.build();
+    }
+
+    private static String validateAndNormalizeName(String n) {
```

Comment:
If you look back a commit it did actually normalize.  I can change it back to be that way if you want.